File: 24. TextJustification.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

vector<string> fullJustify(vector<string>& words, int maxWidth) {
    vector<string> result;
    int i = 0, n = words.size();
    while (i < n) {
        int j = i, len = 0;
        while (j < n && len + (int)words[j].size() + (j - i) <= maxWidth) {
            len += words[j++].size();
        }
        string line;
        int gaps = j - i - 1;
        if (j == n || gaps == 0) {
            for (int k = i; k < j; k++) {
                line += words[k];
                if (k < j - 1) line += ' ';
            }
            while ((int)line.size() < maxWidth) line += ' ';
        } else {
            int spaces = (maxWidth - len) / gaps;
            int extra = (maxWidth - len) % gaps;
            for (int k = i; k < j; k++) {
                line += words[k];
                if (k < j - 1) {
                    for (int s = 0; s < spaces; s++) line += ' ';
                    if (k - i < extra) line += ' ';
                }
            }
        }
        result.push_back(line);
        i = j;
    }
    return result;
}

int main() {
    vector<string> words1 = {"This", "is", "an", "example", "of", "text", "justification."};
    vector<string> res1 = fullJustify(words1, 16);
    for (const string& line : res1) cout << line << "\n";

    cout << "\n";

    vector<string> words2 = {"What", "must", "be", "acknowledgment", "shall", "be"};
    vector<string> res2 = fullJustify(words2, 16);
    for (const string& line : res2) cout << line << "\n";

    return 0;
}

File: 24. TextJustification.java
import java.util.ArrayList;
import java.util.List;

public class TextJustification {

    public static List<String> fullJustify(String[] words, int maxWidth) {
        List<String> result = new ArrayList<>();
        int i = 0, n = words.length;
        while (i < n) {
            int j = i, len = 0;
            while (j < n && len + words[j].length() + (j - i) <= maxWidth) {
                len += words[j++].length();
            }
            StringBuilder line = new StringBuilder();
            int gaps = j - i - 1;
            if (j == n || gaps == 0) {
                for (int k = i; k < j; k++) {
                    line.append(words[k]);
                    if (k < j - 1) line.append(' ');
                }
                while (line.length() < maxWidth) line.append(' ');
            } else {
                int spaces = (maxWidth - len) / gaps;
                int extra = (maxWidth - len) % gaps;
                for (int k = i; k < j; k++) {
                    line.append(words[k]);
                    if (k < j - 1) {
                        for (int s = 0; s < spaces; s++) line.append(' ');
                        if (k - i < extra) line.append(' ');
                    }
                }
            }
            result.add(line.toString());
            i = j;
        }
        return result;
    }

    public static void main(String[] args) {
        System.out.println(fullJustify(new String[]{"This", "is", "an", "example", "of", "text", "justification."}, 16));
        System.out.println(fullJustify(new String[]{"What", "must", "be", "acknowledgment", "shall", "be"}, 16));
    }
}
